### PR TITLE
Incorporate exchange rates in StableSwap operations

### PR DESCRIPTION
**Pull request by 0xIsshiki**
The business logic changes can be verified by comparing any code in the [Curve contract](https://github.com/curvefi/curve-contract/blob/master/contracts/pools/y/StableSwapY.vy) that uses `_stored_rates()` with the code here in `price.rs` and `curve.rs`. All the other changes are test changes or passing stuff around.

This PR only changes existing tests but doesn't add any new specific tests. Will need to do that, but I thought it was worth getting this PR out in this state.

______
This PR was tested by running unit tests, e2e tests, and fuzz tests for a few minutes.

### CODE REVIEW

**Comment by michaelhly** [comment]
Would you mind merging in the latest master so the CI workflows can run? 🙏

File: stable-swap-client/src/fraction.rs [hunk 1]
```diff
@@ -17,6 +17,20 @@ pub struct Fraction {
...
 }
 
+impl Fraction {
+    /// The fraction corresponding to 1/1.
+    pub const ONE: Self = Self {
+        numerator: 1,
+        denominator: 1,
+    };
+
+    /// The fraction corresponding to 0/0.
```
Comment (michaelhly) [inline stable-swap-client/src/fraction.rs:27]: I think we can just use `Fraction::default` for this?

File: stable-swap-math/src/curve.rs [hunk 2]
```diff
@@ -223,24 +235,37 @@ impl StableSwap {
     }
 
     /// Computes the amount of pool tokens to mint after a deposit.
+    #[allow(clippy::too_many_arguments)]
     pub fn compute_mint_amount_for_deposit(
```
Comment (michaelhly) [inline stable-swap-math/src/curve.rs:239]: Same here for backwards compatibility. 

```rs
compute_mint_amount_for_deposit_with_exchange_rates {
    ....
}
```

File: stable-swap-math/src/curve.rs [hunk 3]
```diff
@@ -322,47 +352,60 @@ impl StableSwap {
     ///
     /// 1. Get current D
     /// 2. Solve Eqn against `y_i` for `D - _token_amount`
+    #[allow(clippy::too_many_arguments)]
     pub fn compute_withdraw_one(
```
Comment (michaelhly) [inline stable-swap-math/src/curve.rs:356]: Same as above for backwards compatibility, if possible.

File: stable-swap-math/src/curve.rs [hunk 4]
```diff
@@ -371,18 +414,30 @@ impl StableSwap {
         source_amount: u64,
         swap_source_amount: u64,
         swap_destination_amount: u64,
+        swap_source_exchange_rate: Fraction,
+        swap_destination_exchange_rate: Fraction,
```
Comment (michaelhly) [inline stable-swap-math/src/curve.rs:418]: Backwards compatibility if possible.

**Review by michaelhly — CHANGES_REQUESTED** [review]
Let's try to preserve backwards compatibility in the `stable-swap-math` create when possible.

**Comment by michaelhly** [comment]
Also it might be a good idea to change the Default trait for `Fraction` to be `ONE`?

**Comment by 0xIsshiki** [comment]
I'm actually not sure the default value of `Fraction` should be 1/1. Maybe if the struct were called `ExchangeRate`, then 1/1 would be the first default that people think of. But I think if you were to ask a random programmer on the street what they thought the default value of a `Fraction` should be, a lot would probably say 0/0.

What do you think?

File: stable-swap-client/src/state.rs [hunk 5]
```diff
@@ -60,6 +60,43 @@ pub struct SwapInfo {
     pub token_b_exchange_rate_override: Fraction,
 }
 
+impl SwapInfo {
+    fn exchange_rate_from_override(exchange_rate_override: Fraction) -> Fraction {
+        if exchange_rate_override.numerator == 0 && exchange_rate_override.denominator == 0 {
```
Comment (michaelhly) [inline stable-swap-client/src/state.rs:65]: Can we implement a method on `Fraction` to check if the denominator is `0`?

File: stable-swap-client/src/state.rs [hunk 6]
```diff
@@ -60,6 +60,43 @@ pub struct SwapInfo {
...
+        if key == self.token_a.reserves {
+            return self.exchange_rate_a();
+        }
+        if key == self.token_b.reserves {
+            return self.exchange_rate_b();
+        }
+        Fraction {
+            numerator: 0,
+            denominator: 0,
+        }
```
Comment (michaelhly) [inline stable-swap-client/src/state.rs:96]:
```suggestion
        Fraction::UNDEFINED
```

File: stable-swap-client/src/instruction.rs [hunk 7]
```diff
@@ -632,11 +654,15 @@ pub fn initialize(
     nonce: u8,
     amp_factor: u64,
     fees: Fees,
+    token_a_exchange_rate_override: Fraction,
+    token_b_exchange_rate_override: Fraction,
```
Comment (michaelhly) [inline stable-swap-client/src/instruction.rs:658]: Let's follow up and revert the override setters in `init`. Instead, let us will set these by combining the init instruction with the setter instruction(s) client-side.

**Review by michaelhly — APPROVED** [review]

**Pull request merged**